### PR TITLE
dforce subgraph

### PR DESCRIPTION
**Pull request by tnkrxyz**
This is a work-in-progress PR for the dforce subgraph (mainnet):

- [x] debug subgraph indexing

- [x] verify information correctness

- [x]  incorporate any feedback

Relevant commit starting from f4fbb6fb65068f5dd90c1d16193ded4959fcd3fd (commits prior have been merged)

### CODE REVIEW

File: subgraphs/inverse-finance/src/common/helpers.ts [hunk 1]
```diff
@@ -490,6 +490,24 @@ export function updateRevenue(
 //    - FinancialsDailySnapshot.cumulativeTotalRevenueUSD
 export function updateStablizerFees(fees: BigDecimal, event: ethereum.Event): void {
   let protocol = getOrCreateProtocol();
+<<<<<<< HEAD
```
Comment (this-username-is-taken) [inline subgraphs/inverse-finance/src/common/helpers.ts:493]: merge conflict

**Comment by this-username-is-taken** [comment]
You should start the PR from the latest master

**Comment by this-username-is-taken** [comment]
Can you talk to @0xbe1 to see if this could be integrated into `compound-forks`?

**Comment by 0xbe1** [comment]
@this-username-is-taken i have little context about dforce - is it a compound fork?

**Comment by tnkrxyz** [comment]
> You should start the PR from the latest master

Rebased.

> Can you talk to @0xbe1 to see if this could be integrated into `compound-forks`?

The original codebase of dforce certainly was based on compound. I will look into the compound fork subgraph to see if it makes sense to integrate.

**Comment by tnkrxyz** [comment]
My reading is that dforce's code moved further away from the compound codebase compared to the few protocols currently implemented as compound-forks. A few notable differences:

1. CToken is renamed to iToken (I suppose that we can trick it to still call it cToken)
2. Events have different names/interfaces
      1. It does not have a `ActionPaused` event, instead Borrow, Mint, Redeem, Transfer each has a pause event (MintPaused etc)
      2. MarketListed is renamed to MarketAdded, with `MarketAdded(address,uint256,uint256,uint256,uint256,uint256)` vs `MarketListed(address)`
      3. AccrueInterest is renamed to UpdateInterest with a slight different interface
      4. Mint, Redeem, Borrow, RepayBorrow events all have different number of parameters 
6. Reward emission in dforce is handled by a separate contract RewardDistributor
7. dforce has two stablecoins

I wouldn't say it is impossible to integrate dforce, but the integration will be less clear than other compound-forks, likely requires some tricks (faking iToken as a cToken) and/or duplicating code.

Comments? Suggestions?

**Comment by this-username-is-taken** [comment]
1. I believe there are some compound forks where the cToken is named as xToken that you can use as a reference.
2. I believe the compound forks codebase is designed in a way that it should accommodate minor differences in ABIs
3. I see. You might need a separate handler for that
4. The Venus fork also has a stablecoin which should be relatively easy to add

Why don't you give it a shot and see how it looks like. If it's too messy we can leave it separate

**Comment by 0xbe1** [comment]
@tnkrxyz 

1. yeah just rename iToken to cToken in the abi. I find it an ok hack
2. then just use separate handlers
3. existing compound-fork framework doesn't regulate how you should do reward. you can feel free to plug in dforce's reward mechanism. it is designed to not to worry about reward, and leave it for specific implementations
4. similar to 3, this should be fine.

**Comment by tnkrxyz** [comment]
Thanks for the input! Working on refactoring my mappings as a compound-fork.

**Comment by tnkrxyz** [comment]
@this-username-is-taken @0xbe1 I submitted the first version of dforce subgraph refactored as a compound-fork and hope to run a few things by you.

One major deviation from your suggestion is that, instead of hacking abis, I refactored compound-forks/src/mapping.ts to be more generic - getting needed information from function arguments instead of from the event object.  I am hestitant of hacking abis by renaming events & params in that there is not a good way to document what has been changed & how to convert actual abis to the hacked version. You can see a diff of the changes I made to compound-forks/src/mapping.ts  here: https://github.com/messari/subgraphs/pull/263/commits/1807712ae4d4ade208b8b369f3875ef0c25a0ef2#diff-0064b333bcaf6b3772dd082a031527ff07db0c65c9c2d1055e3b87ee3da3f319 . If you feel this is a sensible change/improvement. I will refactor current compound-forks to to use the new compound-forks/src/mapping.ts

Otherwise, implemenation of the dforce subgraph is in line with your suggestions.

**Comment by 0xbe1** [comment]
@tnkrxyz haven't looked into the exact code change, but your deviation makes a lot of sense. I will take a look in a few days. thank you for the improvement

**Comment by 0xbe1** [comment]
@tnkrxyz I do think the change is an improvement and it would be great if you can lead the refactor.

let me also tag @dmelotik since he is also very familiar with compound forks. Dylan, what do you think?

**Comment by tnkrxyz** [comment]
Great! I will start refactoring existing compound-forks once I get a go-ahead from @dmelotik & start to wrap up the dforce subgraph (dependent on this change).

**Comment by dmelotik** [comment]
@tnkrxyz Sorry I didn't see this sooner. But I think this is a good change, more future-proof and generic. 

lgtm

**Comment by tnkrxyz** [comment]
@dmelotik , @0xbe1 , I refactored the compound-forks following our discussion above in PR #317 . Appreciate it if either of you can help review it.

**Comment by tnkrxyz** [comment]
@this-username-is-taken , the dforce subgraph (mainnet, bsc, optimism; the subgraph on the Arbitrum chain is empty) is ready for review.

File: subgraphs/compound-forks/dforce/README.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+## Hosted Subgraph
+
+- Hosted subgraph: https://thegraph.com/hosted-service/subgraph/tnkrxyz/dforce (mainnet)
+- Hosted subgraph: https://thegraph.com/hosted-service/subgraph/tnkrxyz/dforce-bsc (BSC)
+- Hosted subgraph: https://thegraph.com/hosted-service/subgraph/tnkrxyz/dforce-optimism (Optimism)
+- Hosted subgraph: https://thegraph.com/hosted-service/subgraph/tnkrxyz/dforce-arbitrum (Arbitrum-One, empty subgraph)
+
+## Smart Contracts Interactions
+
+![dforce-lending](https://lh6.googleusercontent.com/ZEc8wddU1Etx-ay_fwnat2dgje8yGcOd8EK1vo5JSxJp3GM7qK5bD-bCvf3zxv4m1j5ClAfoCKiPzWNcqjS_XRjrThyz7vGA-ACalWb2XmwqEwGcLMQLSxVtySU7NPGjxPNOqqoB "dforce-lending")
```
Comment (this-username-is-taken) [inline subgraphs/compound-forks/dforce/README.md:98]: you should put the image under `docs` in the root directory

**Comment by this-username-is-taken** [comment]
Looks good to me! Seems like the arbitrum deployment failed syncing (also the TVL on defillama shows >$100M: https://defillama.com/protocol/dforce). Do you know what the delta is?

**Comment by tnkrxyz** [comment]
BSC is syncing and arbitrum re-deployed - I will look into their TVL & compare with defillama later.

For the mainnet, our TVL ($34M) is closer to what dforce's own reported number of $28M (https://df.dforce.network/#/Markets/mainnet?AssetsType=Lend&currentPool=general), while defillama reports $40M. The difference of $6M between our TVL and dforce's is largely due to the discrepancy in the wstETH deposit (we have $8.3M vs $2.8M; onchain ~2838 ETH ~= $2.8M). The TVL calculation is handled by handler functions in generic compound-forks/src/mapping.ts; I will look into this.

**Comment by tnkrxyz** [comment]
@this-username-is-taken Looks like the discrepancy in TVL (between ours & dforce's own) is due to stale price data. compound-forks subgraph only update underlying token price when a UpdateInterest event is emitted, but there has been none for the iwstETH market since May 04, 2022, thus the underlying token price remains at $2944.39 and leads to a TVL of $8.3M. If the price updated to current price of ~$1000, our TVL will matches dforce's number. Short of using an external price lib, I am not sure there is an easy fix to this.

File: subgraphs/compound-forks/dforce/README.md (hunk 2 above)
Comment (tnkrxyz) [reply subgraphs/compound-forks/dforce/README.md]: fixed.

**Review by 0xbe1 — COMMENTED** [review]
lgtm, can't wait to add to liquidated.live 😆

**Comment by this-username-is-taken** [comment]
> @this-username-is-taken Looks like the discrepancy in TVL (between ours & dforce's own) is due to stale price data. compound-forks subgraph only update underlying token price when a UpdateInterest event is emitted, but there has been none for the iwstETH market since May 04, 2022, thus the underlying token price remains at $2944.39 and leads to a TVL of $8.3M. If the price updated to current price of ~$1000, our TVL will matches dforce's number. Short of using an external price lib, I am not sure there is an easy fix to this.

Why do we need an external price lib to do this? Can we update the prices once per day (e.g. triggered by other events)? You can check out StakeDAO's implementation on price updates.

**Comment by tnkrxyz** [comment]
That's true in general - tie price update to events updated more frequently, but UpdateInterest (or AccrueInterest in generic compound-forks) is already the event that's been updated most frequently (called/emitted before mint, redeem, borrow, repay, liquidate). I cannot think of another event that's emitted more frequently. That's reason I brought up an external price lib as potential solution.

For this particular market (dstWETH), no event was emitted between 5/20 and 6/17. It was emitted in the last couple of days, so the TVL ($28M on ethereum) is now inline with dforce's own number.

**Comment by this-username-is-taken** [comment]
> That's true in general - tie price update to events updated more frequently, but UpdateInterest (or AccrueInterest in generic compound-forks) is already the event that's been updated most frequently (called/emitted before mint, redeem, borrow, repay, liquidate). I cannot think of another event that's emitted more frequently. That's reason I brought up an external price lib as potential solution.
> 
> For this particular market (dstWETH), no event was emitted between 5/20 and 6/17. It was emitted in the last couple of days, so the TVL ($28M on ethereum) is now inline with dforce's own number.
> 
> Also, the price update is implemented in compound-forks/src/mapping.ts, so this is not specific to dforce.

What I'm saying is when `UpdateInterest` gets emitted,

**Comment by this-username-is-taken** [comment]
> That's true in general - tie price update to events updated more frequently, but UpdateInterest (or AccrueInterest in generic compound-forks) is already the event that's been updated most frequently (called/emitted before mint, redeem, borrow, repay, liquidate). I cannot think of another event that's emitted more frequently. That's reason I brought up an external price lib as potential solution.
> 
> For this particular market (dstWETH), no event was emitted between 5/20 and 6/17. It was emitted in the last couple of days, so the TVL ($28M on ethereum) is now inline with dforce's own number.
> 
> Also, the price update is implemented in compound-forks/src/mapping.ts, so this is not specific to dforce.

Can you elaborate on why a external price lib would solve this problem?

**Comment by tnkrxyz** [comment]
>  I think I understand the problem. But I'm not sure why bringing in an external price lib would solve this problem. Can you elaborate on that?

My thinking is that if we can find a price oracle contract external to dforce that provides daily price update event, we can use it to trigger update to TVL (daily, or even hourly if desirable).

**Comment by tnkrxyz** [comment]
@this-username-is-taken dforce bsc finally finished syncing, 7 days after deployment. I also re-organized the directory structure following the lead by @dmelotik in #382. Of course dforce bsc subgraph (https://thegraph.com/hosted-service/subgraph/tnkrxyz/dforce-bsc) is still using schema 1.2.1. I will re-deploy it to refresh the schema once you get a chance to take a look (another 7 days wait).

**Comment by 0xbe1** [comment]
> > I think I understand the problem. But I'm not sure why bringing in an external price lib would solve this problem. Can you elaborate on that?
> 
> My thinking is that if we can find a price oracle contract external to dforce that provides daily price update event, we can use it to trigger update to TVL (daily, or even hourly if desirable).

Interesting, which oracle does that?

File: subgraphs/compound-forks/protocols/dforce/src/mapping.ts [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+    event.params.newLiquidationIncentiveMantissa;
+  _handleNewLiquidationIncentive(protocol, newLiquidationIncentiveMantissa);
+}
+
+// toggle market.isActive based on pause status of mint/redeem/transfer
+export function handleMintPaused(event: MintPaused): void {
+  let marketID = event.params.iToken.toHexString();
+  let market = Market.load(marketID);
+  if (market == null) {
+    log.warning("[handleNewReserveFactor] Market not found: {}", [marketID]);
```
Comment (0xbe1) [inline subgraphs/compound-forks/protocols/dforce/src/mapping.ts:204]: rename handleNewReserveFactor

File: subgraphs/compound-forks/protocols/dforce/src/mapping.ts [hunk 4]
```diff
@@ -0,0 +1,604 @@
...
+  market.save();
+  _marketStatus.save();
+}
+
+// toggle market.isActive based on pause status of mint/redeem/transfer
+export function handleRedeemPaused(event: RedeemPaused): void {
+  let marketID = event.params.iToken.toHexString();
+  let market = Market.load(marketID);
+  if (market == null) {
+    log.warning("[handleNewReserveFactor] Market not found: {}", [marketID]);
```
Comment (0xbe1) [inline subgraphs/compound-forks/protocols/dforce/src/mapping.ts:227]: rename handleNewReserveFactor

File: subgraphs/compound-forks/protocols/dforce/src/mapping.ts [hunk 5]
```diff
@@ -0,0 +1,604 @@
...
+// toggle market.isActive based on pause status of mint/redeem/transfer
+// transfer pause stops transfer of all iTokens
+export function handleTransferPaused(event: TransferPaused): void {
+  let protocol = getOrCreateProtocol();
+  let markets = protocol._marketIDs;
+  for (let i = 0; i < markets.length; i++) {
+    let marketID = markets[i];
+    let market = Market.load(marketID);
+    if (market == null) {
+      log.warning("[handleNewReserveFactor] Market not found: {}", [marketID]);
```
Comment (0xbe1) [inline subgraphs/compound-forks/protocols/dforce/src/mapping.ts:254]: rename handleNewReserveFactor

File: subgraphs/compound-forks/protocols/dforce/src/mapping.ts [hunk 6]
```diff
@@ -0,0 +1,604 @@
...
+
+export function handleBorrowPaused(event: BorrowPaused): void {
+  // toggle market.canBorrowFrom based on BorrowPaused event
+  let marketId = event.params.iToken.toHexString();
+  let market = Market.load(marketId);
+  if (market != null) {
+    market.canBorrowFrom = !event.params.paused;
+    market.save();
+  } else {
+    log.warning("Market {} does not exist.", [marketId]);
```
Comment (0xbe1) [inline subgraphs/compound-forks/protocols/dforce/src/mapping.ts:280]: add "[handleBorrowPaused]" to be consistent

File: subgraphs/compound-forks/schema.graphql [hunk 7]
```diff
@@ -1074,3 +1074,16 @@ type _CircularBuffer @entity {
   " The current calculated number of blocks per day based on calculated block speed "
   blocksPerDay: BigDecimal!
 }
+
+# Helper entity for determining market.isActive status (dforce)
+# market.isActive = false if either mintPaused, redeemPaused, transferPaused = true
+type _MarketStatus @entity {
```
Comment (0xbe1) [inline subgraphs/compound-forks/schema.graphql:1080]: given this is dforce-specific can you name it `_DforceMarketStatus`?

**Review by 0xbe1 — APPROVED** [review]
you also want to include source code of the dforce chart, preferably in draw.io format if it is what you use

**Comment by tnkrxyz** [comment]
> Interesting, which oracle does that?

Not sure the update frequency Chainlink's priceOracle (haven't looked into it and not sure if suitable for our use case). For stETH, it should be more frequent than a month.

File: subgraphs/compound-forks/protocols/dforce/src/mapping.ts (hunk 3 above)
Comment (tnkrxyz) [reply subgraphs/compound-forks/protocols/dforce/src/mapping.ts]: fixed

File: subgraphs/compound-forks/protocols/dforce/src/mapping.ts (hunk 4 above)
Comment (tnkrxyz) [reply subgraphs/compound-forks/protocols/dforce/src/mapping.ts]: fixed

File: subgraphs/compound-forks/protocols/dforce/src/mapping.ts (hunk 5 above)
Comment (tnkrxyz) [reply subgraphs/compound-forks/protocols/dforce/src/mapping.ts]: fixed

File: subgraphs/compound-forks/protocols/dforce/src/mapping.ts (hunk 6 above)
Comment (tnkrxyz) [reply subgraphs/compound-forks/protocols/dforce/src/mapping.ts]: fixed

File: subgraphs/compound-forks/schema.graphql (hunk 7 above)
Comment (tnkrxyz) [reply subgraphs/compound-forks/schema.graphql]: fixed

**Comment by tnkrxyz** [comment]
> you also want to include source code of the dforce chart, preferably in draw.io format if it is what you use

The chart was borrowed from dforce's offical chart: https://lh6.googleusercontent.com/ZEc8wddU1Etx-ay_fwnat2dgje8yGcOd8EK1vo5JSxJp3GM7qK5bD-bCvf3zxv4m1j5ClAfoCKiPzWNcqjS_XRjrThyz7vGA-ACalWb2XmwqEwGcLMQLSxVtySU7NPGjxPNOqqoB. I didn't find the source code.

**Comment by 0xbe1** [comment]
> > you also want to include source code of the dforce chart, preferably in draw.io format if it is what you use
> 
> 
> 
> The chart was borrowed from dforce's offical chart: https://lh6.googleusercontent.com/ZEc8wddU1Etx-ay_fwnat2dgje8yGcOd8EK1vo5JSxJp3GM7qK5bD-bCvf3zxv4m1j5ClAfoCKiPzWNcqjS_XRjrThyz7vGA-ACalWb2XmwqEwGcLMQLSxVtySU7NPGjxPNOqqoB. I didn't find the source code.

Interesting, then link to the website which shows the chart

**Comment by tnkrxyz** [comment]
> > > you also want to include source code of the dforce chart, preferably in draw.io format if it is what you use
> > 
> > 
> > The chart was borrowed from dforce's offical chart: https://lh6.googleusercontent.com/ZEc8wddU1Etx-ay_fwnat2dgje8yGcOd8EK1vo5JSxJp3GM7qK5bD-bCvf3zxv4m1j5ClAfoCKiPzWNcqjS_XRjrThyz7vGA-ACalWb2XmwqEwGcLMQLSxVtySU7NPGjxPNOqqoB. I didn't find the source code.
> 
> Interesting, then link to the website which shows the chart

OK added.

**Comment by tnkrxyz** [comment]
> > Interesting, which oracle does that?
> 
> Not sure the update frequency Chainlink's priceOracle (haven't looked into it and not sure if suitable for our use case). For stETH, it should be more frequent than a month.

For example, this priceOracle feed by ChainLink emits events when stETH price is updated: https://etherscan.io/address/0x37bc7498f4ff12c19678ee8fe19d713b87f6a9e6#events

**Comment by this-username-is-taken** [comment]
Everything looks good. Merging this and addressing the `wsteth` price issue in: https://github.com/messari/subgraphs/pull/429

**Pull request merged**